src/core/ucc_context.c

```diff
@@ -1131,3 +1131,234 @@ ucc_status_t ucc_context_get_attr(ucc_context_t      *context,
+       addresses, etc. likely garbage. fix it */
+    local_memh->tl_h = (ucc_mem_map_tl_t *)ucc_calloc(
+        ctx->n_tl_ctx, sizeof(ucc_mem_map_tl_t), "tl memh");
+    for (i = 0; i < ctx->n_tl_ctx; i++) {
+        tl_lib = ucc_derived_of(ctx->tl_ctx[i]->super.lib, ucc_tl_lib_t);
+        strncpy(local_memh->tl_h[i].tl_name, tls->names[i], UCC_MEM_MAP_TL_NAME_LEN - 1);
+        status = tl_lib->iface->context.mem_map(
+            (const ucc_base_context_t *)ctx->tl_ctx[i], type,
+            params->segments[0].address, params->segments[0].len, local_memh,
+            &local_memh->tl_h[i]);
+        if (status < UCC_ERR_NOT_IMPLEMENTED) {
+            ucc_error("failed to import mem map memh %d", status);
+            return status;
+        }
+    }
+    local_memh->type = type;
+    /* fix context as it will be incorrect on a different system */
+    local_memh->context = ctx;
+    *memh_size          = 0;
+
+    return status;
+}
+
+ucc_status_t ucc_mem_map_export(ucc_context_h         context,
+                                ucc_mem_map_flags_t   flags,
+                                ucc_mem_map_params_t *params, size_t *memh_size,
+                                ucc_mem_map_mem_h *memh)
+{
+    ucc_context_t            *ctx             = (ucc_context_t *)context;
+    ucc_config_names_array_t *tls             = &ctx->all_tls;
+    size_t                    total_pack_size = 0;
+    ucc_mem_map_memh_t       *local_memh;
+    ucc_mem_map_memh_t       *exported_memh;
+    void                    **packed_buffers;
+    ucc_status_t              status;
+    ucc_tl_lib_t             *tl_lib;
+    size_t                    offset;
+    int                       i;
+    ucc_mem_map_type_t        type;
+
+    if (flags == UCC_MEM_MAP_EXPORT) {
+        local_memh = (ucc_mem_map_memh_t *)ucc_calloc(1, sizeof(ucc_mem_map_memh_t),
+                                                      "local memh");
+        if (!local_memh) {
+            ucc_error("failed to allocate a local memory handle");
+            return UCC_ERR_NO_MEMORY;
+        }
+
+        local_memh->tl_h = (ucc_mem_map_tl_t *)ucc_calloc(
+            ctx->n_tl_ctx, sizeof(ucc_mem_map_tl_t), "tl memh");
+        local_memh->address = params->segments[0].address;
+        local_memh->len = params->segments[0].len;
+        type = UCC_MEM_MAP_TYPE_LOCAL;
+    } else {
+        local_memh = *memh;
+        type = UCC_MEM_MAP_TYPE_OFFLOAD_EXPORT;
+    }
+    packed_buffers =
+        (void **)ucc_calloc(ctx->n_tl_ctx, sizeof(void *), "packed buffers");
+
+    /* map all the memory */
+    for (i = 0; i < ctx->n_tl_ctx; i++) {
+        tl_lib = ucc_derived_of(ctx->tl_ctx[i]->super.lib, ucc_tl_lib_t);
+        /* always treat as a local mem handle */
+        status = tl_lib->iface->context.mem_map(
+            (const ucc_base_context_t *)ctx->tl_ctx[i], type,
+            params->segments[0].address, params->segments[0].len, local_memh,
```

Comment:
address and len are allready part of memh, why do we need to pass them separately one more time?

Reply:
fixed